### PR TITLE
wayland: implement screens.

### PR DESCRIPTION
**Pull request by james-lawrence**
- implements the logic to pull screen information as needed.
- refactors some of the wayland connection code out of the
  application startup.

### CODE REVIEW

File: druid-shell/src/backend/wayland/application.rs [hunk 1]
```diff
@@ -180,120 +167,136 @@ impl Application {
...
-                                "internal: wayland should always use new IDs"
-                            );
-                            // Defer setting up the pointer/keyboard event handling until we've
-                            // finished constructing the `Application`. That way we can pass it as a
-                            // parameter.
-                        }
+
+        let dispatcher = display::Dispatcher::default();
+
+        display::GlobalEventDispatch::subscribe(&dispatcher, move |event: &'_ wl::GlobalEvent, registry: &'_ wl::Attached<wl_registry::WlRegistry>, _ctx: &'_ wl::DispatchData| {
```
Comment (james-lawrence) [inline druid-shell/src/backend/wayland/application.rs:173]: looking to pull this entire logic into the outputs module in next PR where I'll be fixing a layershell reinitialization bug.

File: druid-shell/src/backend/wayland/screen.rs [hunk 2]
```diff
@@ -12,11 +12,41 @@
...
+    let monitors: Vec<Monitor> = metas
+        .iter()
+        .map(|m| {
+            let rect = kurbo::Rect::from_points(
+                (m.position.x as f64, m.position.y as f64),
+                (
+                    (m.position.x + m.logical.width) as f64,
+                    (m.position.y + m.logical.height) as f64,
+                ),
+            );
```
Comment (Maan2003) [inline druid-shell/src/backend/wayland/screen.rs:34]: maybe?
```suggestion
            let rect = kurbo::Rect::from_origin_size(
                (m.position.x as f64, m.position.y as f64),
                (m.logical.width as f64, m.logical.height as f64),
            );
```

File: druid-shell/src/backend/wayland/display.rs [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        dispatcher,
+    });
+
+    return Ok(env);
+}
+
+pub(super) fn global() -> Result<std::sync::Arc<Environment>, error::Error> {
+    use lazy_static::lazy_static;
+    lazy_static! {
+        static ref _GLOBAL: std::sync::Mutex<Option<std::cell::RefCell<std::sync::Arc<Environment>>>> =
```
Comment (Maan2003) [inline druid-shell/src/backend/wayland/display.rs:191]: `Mutex<Option<RefCell<Arc<>>>>` feels very weird, `Mutex<Option<Arc<>>>` should just work.

File: druid-shell/src/backend/wayland/display.rs [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+pub(super) struct Environment {
+    pub(super) display: wlc::Display,
+    pub(super) registry: wlc::GlobalManager,
+    pub(super) xdg_base: wlc::Main<xdg_wm_base::XdgWmBase>,
+    pub(super) queue: std::rc::Rc<std::cell::RefCell<wlc::EventQueue>>,
+    dispatcher: std::sync::Arc<Dispatcher>,
+}
+
+unsafe impl Sync for Environment {}
+unsafe impl Send for Environment {}
```
Comment (Maan2003) [inline druid-shell/src/backend/wayland/display.rs:124]: Please explain these in a comment :grimacing:

File: druid-shell/src/backend/wayland/display.rs [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+}
+
+pub trait GlobalEventConsumer {
+    fn consume<'a>(
+        &self,
+        event: &'a wlc::GlobalEvent,
+        registry: &'a wlc::Attached<wl_registry::WlRegistry>,
+        ctx: &'a wlc::DispatchData,
+    );
+}
```
Comment (Maan2003) [inline druid-shell/src/backend/wayland/display.rs:71]: you don't need these lifetimes btw

**Review by Maan2003 — APPROVED** [review]
Nice refactor! Just some minor nits

File: druid-shell/src/backend/wayland/display.rs (hunk 3 above)
Comment (james-lawrence) [reply druid-shell/src/backend/wayland/display.rs:191]: i remember this particular piece being a PITA to determine. will take another look now that the code is working and stabilized though.

Comment (Maan2003) [reply druid-shell/src/backend/wayland/display.rs:191]: also if you want a "global storage" you can create these in `Application::new()` and store it in `Application`, then you can access the global data using `druid_shell::Application::global().0.data`

Comment (james-lawrence) [reply druid-shell/src/backend/wayland/display.rs]: indeed, thought about it, dislike reaching outside the module and making that particular assumption for this. either screens is global and doesn't depending on applications or its not and it should take a reference of some sort in its function args.

File: druid-shell/src/backend/wayland/display.rs [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+
+pub(super) fn count<'a>(reg: &'a wlc::GlobalManager, i: &str) -> u32 {
+    reg.list()
+        .iter()
+        .fold(0, |total, (_id, name, _version)| -> u32 {
+            if name == i {
+                return total + 1;
+            }
+            return total;
+        })
```
Comment (Maan2003) [inline druid-shell/src/backend/wayland/display.rs:228]:
```suggestion
        .filter(|_, name, _) name == i)
        .count()
```

**Pull request merged**